pkg/planner/core/memtable_infoschema_extractor.go

```diff
@@ -0,0 +1,493 @@
+}
+
+// InfoSchemaSchemataExtractor is the predicate extractor for information_schema.schemata.
+type InfoSchemaSchemataExtractor struct {
+	InfoSchemaBaseExtractor
+}
+
+// ListSchemas lists related schemas from predicate.
+// If no schema found in predicate, it return all schemas.
+func (e *InfoSchemaSchemataExtractor) ListSchemas(is infoschema.InfoSchema) []model.CIStr {
+	return e.listSchemas(is, _schemaName)
+}
+
+func (e *InfoSchemaBaseExtractor) listSchemas(is infoschema.InfoSchema, schemaCol string) []model.CIStr {
+	schemas := e.getSchemaObjectNames(schemaCol)
+	if len(schemas) == 0 {
+		ret := is.AllSchemaNames()
+		slices.SortFunc(ret, func(a, b model.CIStr) int {
+			return strings.Compare(a.L, b.L)
+		})
+		return ret
+	}
+	ret := schemas[:0]
+	for _, s := range schemas {
+		if n, ok := is.SchemaByName(s); ok {
+			ret = append(ret, n.Name)
+		}
+	}
+	return ret
+}
+
+func findNameAndAppendToTableMap(
+	ctx context.Context,
+	is infoschema.InfoSchema,
+	schema model.CIStr,
+	tableNames []model.CIStr,
+	tables map[int64]*model.TableInfo,
+) error {
+	for _, n := range tableNames {
+		tbl, err := is.TableByName(ctx, schema, n)
+		if err != nil {
+			if terror.ErrorEqual(err, infoschema.ErrTableNotExists) {
+				continue
+			}
+			return errors.Trace(err)
+		}
+		tblInfo := tbl.Meta()
+		tables[tblInfo.ID] = tblInfo
+	}
+	return nil
+}
+
+// findTablesByID finds tables by table IDs and append them to table map.
+func findTablesByID(
+	is infoschema.InfoSchema,
+	tableIDs []model.CIStr,
+	tableNames []model.CIStr,
+	tables map[int64]*model.TableInfo,
+) {
+	for _, tid := range parseIDs(tableIDs) {
+		tbl, ok := is.TableByID(tid)
+		if !ok {
+			continue
+		}
+		if len(tableNames) > 0 && containInTableNames(tableNames, tbl) {
+			continue
+		}
+		tblInfo := tbl.Meta()
+		tables[tblInfo.ID] = tblInfo
+	}
+}
+
+// findTablesByPartID finds tables by partition IDs and append them to table map.
+func findTablesByPartID(
+	is infoschema.InfoSchema,
+	partIDs []model.CIStr,
+	tableNames []model.CIStr,
+	tables map[int64]*model.TableInfo,
+) {
+	for _, pid := range parseIDs(partIDs) {
+		tbl, _, _ := is.FindTableByPartitionID(pid)
+		if tbl == nil {
+			continue
+		}
+		if len(tableNames) > 0 && containInTableNames(tableNames, tbl) {
+			continue
+		}
+		tblInfo := tbl.Meta()
+		tables[tblInfo.ID] = tblInfo
+	}
+}
+
+func containInTableNames(tableNames []model.CIStr, tbl table.Table) bool {
+	for _, n := range tableNames {
+		if tbl.Meta().Name.L == n.L {
+			return true
+		}
+	}
+	return false
+}
+
+func findTableAndSchemaByName(
+	ctx context.Context,
+	is infoschema.InfoSchema,
+	schemas []model.CIStr,
+	tableNames []model.CIStr,
+) ([]model.CIStr, []*model.TableInfo, error) {
+	type schemaAndTable struct {
+		schema model.CIStr
+		table  *model.TableInfo
+	}
+	tableMap := make(map[int64]schemaAndTable, len(tableNames))
+	for _, n := range tableNames {
+		for _, s := range schemas {
+			tbl, err := is.TableByName(ctx, s, n)
```

Comment:
I see at least in infoschema v2 when we get the table we already know its db (`tableItem` has associaten all related information together). So there's a potential optimization to skip the `TableByName`.

Reply:
Maybe we can add more methods for infoschema v2 later.